fix: lower the minimum Dart SDK version to 2.14

dshukertjr opened the pull request:
Remove all the extended enum use in the library to lower the minimum SDK version to 2.12.

dshukertjr commented:
It seems like for this one the minimum SDK version can be 2.12.0 that is the minimum SDK version of all the dependencies.

Vinzent03 commented on lib/src/realtime_channel.dart:
```diff
@@ -61,11 +61,9 @@ class ChannelFilter {
-  rateLimited;
+enum ChannelResponse { ok, timedOut, rateLimited }
 
+extension Name on ChannelResponse {
```
I would rename the extension, because it's exported and may cause a conflict with another extension in the user's project.

dshukertjr replied:
Come to think about it, I don't think we need to actually convert `ChannelResponse` to a string at all, so we can get rid of it all together.

Vinzent03 commented on lib/src/realtime_channel.dart:
```diff
@@ -78,11 +76,9 @@ enum ChannelResponse {
-  presence;
+enum RealtimeListenTypes { postgresChanges, broadcast, presence }
 
+extension ToType on RealtimeListenTypes {
```
Here as well

dshukertjr replied:
Thanks for picking up on these! I think this one we can just hide it and not make it accessible to the users!

Vinzent03 commented on lib/src/constants.dart:
```diff
@@ -23,17 +23,19 @@ enum ChannelEvents {
+  static ChannelEvents fromType(String type) {
     for (ChannelEvents enumVariant in ChannelEvents.values) {
-      if (enumVariant.name == type) return enumVariant;
+      if (enumVariant.name == type || enumVariant.eventName() == type) {
```
Is `enumVariant.name` usable on Dart versions not supporting this? Like, doesn't it throw an `MethodNotFoundException` on Dart 2.12?

dshukertjr replied:
Great call! Let me add a test to check the code against v2.12.0 of Dart!

dshukertjr replied:
And you're right. `.name` was only introduced in 2.15.0. 
https://github.com/dart-lang/sdk/blob/main/CHANGELOG.md#dartcore-4

Will fix by adding a name extension method.

Also, Github Actions with specifying the Dart version doesn't seem to be working
https://github.com/dart-lang/setup-dart/issues/61

Vinzent03 commented:
Do we want to include the 2.12.0 job, although the job doesn't work? I think we should add them, when the issue resolved is.

Vinzent03 commented:
No wait, I just looked into the actions log. The test fails, because the lint requires Dart 2.17. Doesn't this affect all dart libraries, so the Dart version lowering was useless? Unless we downgrade `lints` as well?

dshukertjr commented:
Lowering SDK is pretty tough huh? At this point I'm regretting lowering the minimum SDK 😂

dshukertjr commented:
Change of plan. Going to make the minimum SDK to 2.14.0 from 2.12.0, because it allows us to create extension method `name` to all Enums at once.

```dart
extension EnumName on Enum {
  String get name {
    return toString().split('.').last;
  }
}
```

Vinzent03 commented:
@dshukertjr I don't know exactly why you choose 2.14 now. My previous comment is the same for 2.12 and 2.14, right? Is it not possible to add an extension to the `Enum` type in 2.12? I'm fine with 2.14, but I want to understand your reason.

dshukertjr commented:
@Vinzent03 
Sorry for the lack of details in my explanation. With 2.12, it seemed like creating an extension method on `Enum` class is not allowed, and we would have to target each individual enums to add `name` method.

This is allowed in 2.14.0, but not in 2.12.0
```dart
extension EnumName on Enum {
  String get name {
    return toString().split('.').last;
  }
}
```

This is allowed in 2.12.0, because it's targeting a specific Enum `ChannelEvents`
```dart
extension EnumName on ChannelEvents {
  String get name {
    return toString().split('.').last;
  }
}
```

You can find the failing action here:
https://github.com/supabase-community/realtime-dart/actions/runs/3265055768/jobs/5366663656
